Run the tests with the cache off

And also fix the current set of failures, of which there are quite a few. https://github.com/sympy/sympy/pull/12908 fixes some. It looks like there are more similar recursion issues in assumptions. 

If you don't know, you can run the tests with the cache off using `./bin/test -C`. Or `SYMPY_USE_CACHE="no" ./bin/test`. And be warned that things are *way* slower without the cache.

There is a really old issue about cache test failures https://github.com/sympy/sympy/issues/5594, but I'm guessing anything in that issue is irrelevant by now, so I'm closing it. 

The hard question is, how can we run the test suite automatically with the cache off. It takes a really long time to finish, so adding it to Travis, or even a separate service like Circle CI would really slow things down. But on the other hand it's annoying that these failures pile up because no one tries turning the cache off.

I just ran the core tests and got one failure which can be reproduced with:
```console
$ SYMPY_USE_CACHE="no" pytest sympy/core/ -k test_Mul_is_even_odd --tb=short
==================================================================== test session starts =====================================================================
platform darwin -- Python 3.5.7, pytest-4.3.1, py-1.8.0, pluggy-0.9.0
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/Users/enojb/current/sympy/sympy/.hypothesis/examples')
architecture: 64-bit
cache:        no
ground types: python 

rootdir: /Users/enojb/current/sympy/sympy, inifile: pytest.ini
plugins: xdist-1.27.0, instafail-0.4.1, forked-1.0.2, doctestplus-0.3.0, cov-2.7.1, hypothesis-4.32.3
collected 1818 items / 1817 deselected / 1 selected                                                                                                          

sympy/core/tests/test_arit.py F                                                                                                                        [100%]

========================================================================== FAILURES ==========================================================================
____________________________________________________________________ test_Mul_is_even_odd ____________________________________________________________________
sympy/core/tests/test_arit.py:443: in test_Mul_is_even_odd
    assert (k/2).is_integer is False
E   assert None is False
E    +  where None = (k / 2).is_integer
                                                                       DO *NOT* COMMIT!                                                                       
========================================================= 1 failed, 1817 deselected in 2.31 seconds ==========================================================
```

Running the tests in functions I found this one but I can only reproduce this intermittently (it seems to happen more reliably in Python 3.8 than 3.5):
```console
$ SYMPY_USE_CACHE="no" pytest sympy/concrete/tests/test_sums_products.py -k test_issue_2787
==================================================================== test session starts =====================================================================
platform darwin -- Python 3.8.0b1, pytest-4.6.3, py-1.8.0, pluggy-0.12.0
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/Users/enojb/current/sympy/sympy/.hypothesis/examples')
architecture: 64-bit
cache:        no
ground types: gmpy 1.17

rootdir: /Users/enojb/current/sympy/sympy, inifile: pytest.ini
plugins: xdist-1.29.0, forked-1.0.2, instafail-0.4.1, hypothesis-4.28.2
collected 80 items / 79 deselected / 1 selected                                                                                                              

sympy/concrete/tests/test_sums_products.py F                                                                                                           [100%]

========================================================================== FAILURES ==========================================================================
______________________________________________________________________ test_issue_2787 _______________________________________________________________________

    def test_issue_2787():
        n, k = symbols('n k', positive=True, integer=True)
        p = symbols('p', positive=True)
        binomial_dist = binomial(n, k)*p**k*(1 - p)**(n - k)
        s = Sum(binomial_dist*k, (k, 0, n))
        res = s.doit().simplify()
>       assert res == Piecewise(
            (n*p, p/Abs(p - 1) <= 1),
            ((-p + 1)**n*Sum(k*p**k*(-p + 1)**(-k)*binomial(n, k), (k, 0, n)),
            True))
E       assert Piecewise((n*p, (p/Abs(p - 1) <= 1) & ((n <= 1) | (p/Abs(p - 1) <= 1))), ((1 - p)**n*Sum(k*p**k*(1 - p)**(-k)*binomial(n, k), (k, 0, n)), True)) == Piecewise((n*p, p/Abs(p - 1) <= 1), ((1 - p)**n*Sum(k*p**k*(1 - p)**(-k)*binomial(n, k), (k, 0, n)), True))
E        +  where Piecewise((n*p, p/Abs(p - 1) <= 1), ((1 - p)**n*Sum(k*p**k*(1 - p)**(-k)*binomial(n, k), (k, 0, n)), True)) = Piecewise((n*p, p/Abs(p - 1) <= 1), ((1 - p)**n*Sum(k*p**k*(1 - p)**(-k)*binomial(n, k), (k, 0, n)), True))

sympy/concrete/tests/test_sums_products.py:930: AssertionError
                                                                       DO *NOT* COMMIT!                                                                       
========================================================== 1 failed, 79 deselected in 99.57 seconds ==========================================================
```
The expressions being compared are:
```julia
In [6]: Piecewise((n*p, p/Abs(p - 1) <= 1), ((1 - p)**n*Sum(k*p**k*(1 - p)**(-k)*binomial(n, k), (k, 0, n)), True))                                           
Out[6]: 
⎧                                          p       
⎪               n⋅p                 for ─────── ≤ 1
⎪                                       │p - 1│    
⎪                                                  
⎪           n                                      
⎪          ____                                    
⎪          ╲                                       
⎨           ╲                                      
⎪       n    ╲     k        -k ⎛n⎞                 
⎪(1 - p) ⋅   ╱  k⋅p ⋅(1 - p)  ⋅⎜ ⎟     otherwise   
⎪           ╱                  ⎝k⎠                 
⎪          ╱                                       
⎪          ‾‾‾‾                                    
⎪         k = 0                                    
⎩                                                  

In [7]: res                                                                                                                                                   
Out[7]: 
⎧                                          p          ⎛           p       ⎞
⎪               n⋅p                 for ─────── ≤ 1 ∧ ⎜n ≤ 1 ∨ ─────── ≤ 1⎟
⎪                                       │p - 1│       ⎝        │p - 1│    ⎠
⎪                                                                          
⎪           n                                                              
⎪          ____                                                            
⎪          ╲                                                               
⎨           ╲                                                              
⎪       n    ╲     k        -k ⎛n⎞                                         
⎪(1 - p) ⋅   ╱  k⋅p ⋅(1 - p)  ⋅⎜ ⎟                 otherwise               
⎪           ╱                  ⎝k⎠                                         
⎪          ╱                                                               
⎪          ‾‾‾‾                                                            
⎪         k = 0                                                            
⎩                
```

We can see errors from running without cache very quickly when running the series tests:
```
SYMPY_USE_CACHE="no" pytest sympy/series -m 'not slow' -n4 -d --instafail
```
E.g.:
```
______________________________________________________________________ test_basic5 _________________________________________________________________________
[gw0] darwin -- Python 3.8.0 /Users/enojb/current/sympy/38venv/bin/python3.8

    def test_basic5():
        class my(Function):
            @classmethod
            def eval(cls, arg):
                if arg is S.Infinity:
                    return S.NaN
>       assert limit(my(x), x, oo) == Limit(my(x), x, oo)

sympy/series/tests/test_limits.py:110: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
sympy/series/limits.py:71: in limit
    return Limit(e, z, z0, dir).doit(deep=False)
sympy/series/limits.py:261: in doit
    r = gruntz(e, z, z0, dir)
sympy/series/gruntz.py:646: in gruntz
    r = limitinf(e, z)
sympy/series/gruntz.py:426: in limitinf
    c0, e0 = mrv_leadterm(e, x)
sympy/series/gruntz.py:511: in mrv_leadterm
    series = calculate_series(f, w, logx=logw)
sympy/series/gruntz.py:464: in calculate_series
    for t in e.lseries(x, logx=logx):
sympy/core/expr.py:2962: in yield_lseries
    for si in s:
sympy/core/expr.py:3174: in _eval_lseries
    series = self._eval_nseries(x, n=n, logx=logx)
sympy/core/function.py:671: in _eval_nseries
    args0 = [t.limit(x, 0) for t in args]
sympy/core/function.py:671: in <listcomp>
    args0 = [t.limit(x, 0) for t in args]
sympy/core/expr.py:3284: in limit
    return limit(self, x, xlim, dir)
sympy/series/limits.py:71: in limit
    return Limit(e, z, z0, dir).doit(deep=False)
sympy/series/limits.py:261: in doit
    r = gruntz(e, z, z0, dir)
sympy/series/gruntz.py:656: in gruntz
    r = limitinf(e0, z)
sympy/series/gruntz.py:426: in limitinf
    c0, e0 = mrv_leadterm(e, x)
sympy/series/gruntz.py:511: in mrv_leadterm
    series = calculate_series(f, w, logx=logw)
sympy/series/gruntz.py:470: in calculate_series
    if t.simplify():
sympy/core/expr.py:3569: in simplify
    return simplify(self, **kwargs)
sympy/simplify/simplify.py:599: in simplify
    expr = Mul(*powsimp(expr).as_content_primitive())
sympy/simplify/powsimp.py:121: in powsimp
    return recurse(expr*_y, deep=False)/_y
sympy/core/decorators.py:91: in __sympifyit_wrapper
    return func(a, b)
sympy/core/decorators.py:129: in binary_op_wrapper
    return func(self, other)
sympy/core/expr.py:233: in __div__
    return Mul(self, Pow(other, S.NegativeOne))
sympy/core/operations.py:47: in __new__
    c_part, nc_part, order_symbols = cls.flatten(args)
sympy/core/mul.py:199: in flatten
    if not a.is_zero and a.is_Rational:
sympy/core/assumptions.py:266: in getit
    return _ask(fact, self)
sympy/core/assumptions.py:321: in _ask
    _ask(pk, obj)
sympy/core/assumptions.py:309: in _ask
    a = evaluate(obj)
sympy/core/mul.py:1249: in _eval_is_integer
    is_rational = self.is_rational
sympy/core/assumptions.py:266: in getit
    return _ask(fact, self)
sympy/core/assumptions.py:309: in _ask
    a = evaluate(obj)
sympy/core/mul.py:1218: in _eval_is_rational
    r = _fuzzy_group((a.is_rational for a in self.args), quick_exit=True)
sympy/core/logic.py:78: in _fuzzy_group
    for a in args:
sympy/core/mul.py:1218: in <genexpr>
    r = _fuzzy_group((a.is_rational for a in self.args), quick_exit=True)
sympy/core/assumptions.py:266: in getit
    return _ask(fact, self)
sympy/core/assumptions.py:321: in _ask
    _ask(pk, obj)
sympy/core/assumptions.py:321: in _ask
    _ask(pk, obj)
sympy/core/assumptions.py:321: in _ask
    _ask(pk, obj)
sympy/core/assumptions.py:321: in _ask
    _ask(pk, obj)
sympy/core/assumptions.py:309: in _ask
    a = evaluate(obj)
sympy/core/power.py:1276: in _eval_is_algebraic
    if self.base.is_zero or _is_one(self.base):
sympy/core/power.py:1271: in _is_one
    return (expr - 1).is_zero
sympy/core/assumptions.py:266: in getit
    return _ask(fact, self)
sympy/core/assumptions.py:321: in _ask
    _ask(pk, obj)
sympy/core/assumptions.py:321: in _ask
    _ask(pk, obj)
sympy/core/assumptions.py:321: in _ask
    _ask(pk, obj)
sympy/core/assumptions.py:321: in _ask
    _ask(pk, obj)
sympy/core/assumptions.py:309: in _ask
    a = evaluate(obj)
sympy/core/add.py:545: in _eval_is_imaginary
    if b.is_zero:
sympy/core/assumptions.py:266: in getit
    return _ask(fact, self)
E   RecursionError: maximum recursion depth exceeded while calling a Python object
!!! Recursion detected (same locals & position)

sympy/series/tests/test_limits.py F
```

I can no longer reproduce the problem in sets. In series there is one test that uses `is` when it should be `==`. I've fixed that in #17704.

There is another failure in series. This test gives infinite recursion:
```console
$ SYMPY_USE_CACHE="no" pytest sympy/series/tests/test_limits.py -k test_issue_13575
```
Some `(erf(I)**2).is_imaginary` ends up creating a new expression of the same form and trying to ask if that is imaginary. With cacheing this works okay because the newly evaluated expression is just the old expression which will have `imaginary=None` in its `_assumptions` dict due to this line:
https://github.com/sympy/sympy/blob/4fc50841e38205561523e0a9d0e5da173988b04a/sympy/core/assumptions.py#L299-L301

I think that disabling that line can expose lots of bugs like this. Maybe that should be opened as a separate issue though.

<details><summary>Click to see full output</summary>

```
$ SYMPY_USE_CACHE="no" pytest sympy/series/tests/test_limits.py -k test_issue_13575
==================================================================== test session starts =====================================================================
platform darwin -- Python 3.5.7, pytest-4.3.1, py-1.8.0, pluggy-0.9.0
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/Users/enojb/current/sympy/sympy/.hypothesis/examples')
architecture: 64-bit
cache:        no
ground types: python 

rootdir: /Users/enojb/current/sympy/sympy, inifile: pytest.ini
plugins: xdist-1.27.0, instafail-0.4.1, forked-1.0.2, doctestplus-0.3.0, cov-2.7.1, hypothesis-4.32.3
collected 70 items / 69 deselected / 1 selected                                                                                                              

sympy/series/tests/test_limits.py F                                                                                                                    [100%]

========================================================================== FAILURES ==========================================================================
______________________________________________________________________ test_issue_13575 ______________________________________________________________________

self = (1 - erf(I))**2 + 2*erf(I)

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'infinite'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = (1 - erf(I))**2 + 2*erf(I)

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'extended_positive'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = (1 - erf(I))**2 + 2*erf(I)

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'extended_real'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = (1 - erf(I))**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'extended_real'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = (1 - erf(I))/sqrt(1 - erf(I)**2)

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'negative'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = (1 - erf(I))/sqrt(1 - erf(I)**2)

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'finite'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = 1/sqrt(1 - erf(I)**2)

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'finite'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = 1 - erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'zero'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'extended_real'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = -erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

self = erf(I)**2

    def getit(self):
        try:
>           return self._assumptions[fact]
E           KeyError: 'imaginary'

sympy/core/assumptions.py:262: KeyError

During handling of the above exception, another exception occurred:

    def test_issue_13575():
>       result = limit(acos(erfi(x)), x, 1)

sympy/series/tests/test_limits.py:545: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
sympy/series/limits.py:71: in limit
    return Limit(e, z, z0, dir).doit(deep=False)
sympy/series/limits.py:261: in doit
    r = gruntz(e, z, z0, dir)
sympy/series/gruntz.py:656: in gruntz
    r = limitinf(e0, z)
sympy/series/gruntz.py:426: in limitinf
    c0, e0 = mrv_leadterm(e, x)
sympy/series/gruntz.py:511: in mrv_leadterm
    series = calculate_series(f, w, logx=logw)
sympy/series/gruntz.py:464: in calculate_series
    for t in e.lseries(x, logx=logx):
sympy/core/expr.py:2962: in yield_lseries
    for si in s:
sympy/core/expr.py:3174: in _eval_lseries
    series = self._eval_nseries(x, n=n, logx=logx)
sympy/functions/elementary/trigonometric.py:2300: in _eval_nseries
    return self._eval_rewrite_as_log(self.args[0])._eval_nseries(x, n, logx)
sympy/core/add.py:407: in _eval_nseries
    terms = [t.nseries(x, n=n, logx=logx) for t in self.args]
sympy/core/add.py:407: in <listcomp>
    terms = [t.nseries(x, n=n, logx=logx) for t in self.args]
sympy/core/expr.py:3261: in nseries
    return self._eval_nseries(x, n=n, logx=logx)
sympy/core/mul.py:1739: in _eval_nseries
    terms = [t.nseries(x, n=n, logx=logx) for t in self.args]
sympy/core/mul.py:1739: in <listcomp>
    terms = [t.nseries(x, n=n, logx=logx) for t in self.args]
sympy/core/expr.py:3261: in nseries
    return self._eval_nseries(x, n=n, logx=logx)
sympy/functions/elementary/exponential.py:893: in _eval_nseries
    s = self.args[0].nseries(x, n=n, logx=logx)
sympy/core/expr.py:3261: in nseries
    return self._eval_nseries(x, n=n, logx=logx)
sympy/core/add.py:407: in _eval_nseries
    terms = [t.nseries(x, n=n, logx=logx) for t in self.args]
sympy/core/add.py:407: in <listcomp>
    terms = [t.nseries(x, n=n, logx=logx) for t in self.args]
sympy/core/expr.py:3261: in nseries
    return self._eval_nseries(x, n=n, logx=logx)
sympy/core/power.py:1537: in _eval_nseries
    if (b0 is S.Zero or b0.is_infinite):
sympy/core/assumptions.py:266: in getit
    return _ask(fact, self)
sympy/core/assumptions.py:321: in _ask
    _ask(pk, obj)
sympy/core/assumptions.py:309: in _ask
    a = evaluate(obj)
sympy/core/expr.py:880: in _eval_is_positive
    extended_positive = self.is_extended_positive
sympy/core/assumptions.py:266: in getit
    return _ask(fact, self)
sympy/core/assumptions.py:309: in _ask
    a = evaluate(obj)
sympy/core/add.py:609: in _eval_is_extended_positive
    return super(Add, self)._eval_is_extended_positive()
sympy/core/expr.py:900: in _eval_is_extended_positive
    if self.is_extended_real is False:
sympy/core/assumptions.py:266: in getit
    return _ask(fact, self)
sympy/core/assumptions.py:309: in _ask
    a = evaluate(obj)
sympy/core/add.py:509: in <lambda>
    (a.is_extended_real for a in self.args), quick_exit=True)
sympy/core/logic.py:78: in _fuzzy_group
    for a in args:
sympy/core/add.py:509: in <genexpr>
    (a.is_extended_real for a in self.args), quick_exit=True)
sympy/core/assumptions.py:266: in getit
    return _ask(fact, self)
sympy/core/assumptions.py:309: in _ask
    a = evaluate(obj)
sympy/core/power.py:627: in _eval_is_extended_real
    i = arg(self.base)*self.exp/S.Pi
sympy/core/function.py:473: in __new__
    result = super(Function, cls).__new__(cls, *args, **options)
sympy/core/function.py:288: in __new__
    evaluated = cls.eval(*args)
sympy/functions/elementary/complexes.py:673: in eval
    rv = atan2(y, x)
sympy/core/function.py:473: in __new__
    result = super(Function, cls).__new__(cls, *args, **options)
sympy/core/function.py:288: in __new__
    evaluated = cls.eval(*args)
sympy/functions/elementary/trigonometric.py:3074: in eval
    (x + S.ImaginaryUnit*y)/sqrt(x**2 + y**2))
sympy/core/function.py:473: in __new__
    result = super(Function, cls).__new__(cls, *args, **options)
sympy/core/function.py:288: in __new__
    evaluated = cls.eval(*args)
sympy/functions/elementary/exponential.py:641: in eval
    if arg.is_negative:
sympy/core/assumptions.py:266: in getit
    return _ask(fact, self)
sympy/core/assumptions.py:309: in _ask
    a = evaluate(obj)
sympy/core/expr.py:887: in _eval_is_negative
    finite = self.is_finite
sympy/core/assumptions.py:266: in getit
    return _ask(fact, self)
sympy/core/assumptions.py:309: in _ask
    a = evaluate(obj)
sympy/core/mul.py:1203: in _eval_is_finite
    if all(a.is_finite for a in self.args):
sympy/core/mul.py:1203: in <genexpr>
    if all(a.is_finite for a in self.args):
sympy/core/assumptions.py:266: in getit
    return _ask(fact, self)
sympy/core/assumptions.py:309: in _ask
    a = evaluate(obj)
sympy/core/power.py:683: in _eval_is_finite
    if self.base.is_zero:
sympy/core/assumptions.py:266: in getit
    return _ask(fact, self)
sympy/core/assumptions.py:309: in _ask
    a = evaluate(obj)
sympy/core/add.py:560: in _eval_is_zero
    if a.is_extended_real:
sympy/core/assumptions.py:266: in getit
    return _ask(fact, self)
sympy/core/assumptions.py:309: in _ask
    a = evaluate(obj)
sympy/core/mul.py:1266: in _eval_is_extended_real
    return self._eval_real_imag(True)
sympy/core/mul.py:1275: in _eval_real_imag
    elif t.is_imaginary:  # I
sympy/core/assumptions.py:266: in getit
    return _ask(fact, self)
sympy/core/assumptions.py:309: in _ask
    a = evaluate(obj)
sympy/core/power.py:664: in _eval_is_imaginary
    i = arg(self.base)*self.exp/S.Pi
sympy/core/function.py:473: in __new__
    result = super(Function, cls).__new__(cls, *args, **options)
sympy/core/function.py:288: in __new__
    evaluated = cls.eval(*args)
sympy/functions/elementary/complexes.py:673: in eval
    rv = atan2(y, x)
sympy/core/function.py:473: in __new__
    result = super(Function, cls).__new__(cls, *args, **options)
sympy/core/function.py:288: in __new__
    evaluated = cls.eval(*args)
sympy/functions/elementary/trigonometric.py:3074: in eval
    (x + S.ImaginaryUnit*y)/sqrt(x**2 + y**2))
sympy/functions/elementary/miscellaneous.py:128: in sqrt
    return Pow(arg, S.Half, evaluate=evaluate)
sympy/core/power.py:300: in __new__
    obj = b._eval_power(e)
sympy/core/mul.py:648: in _eval_power
    if b.is_imaginary:
sympy/core/assumptions.py:266: in getit
    return _ask(fact, self)
sympy/core/assumptions.py:309: in _ask
    a = evaluate(obj)
sympy/core/mul.py:1315: in _eval_is_imaginary
    return self._eval_real_imag(False)
sympy/core/mul.py:1275: in _eval_real_imag
    elif t.is_imaginary:  # I
sympy/core/assumptions.py:266: in getit
    return _ask(fact, self)
E   RecursionError: maximum recursion depth exceeded in __instancecheck__
!!! Recursion detected (same locals & position)
                                                                       DO *NOT* COMMIT!                                                                       
========================================================== 1 failed, 69 deselected in 11.01 seconds ==========================================================

```

</details>

There is one failure in vector:
```console
$ SYMPY_USE_CACHE="no" pytest sympy/vector --instafail -k test_dyadic --tb=short
==================================================================== test session starts =====================================================================
platform darwin -- Python 3.5.7, pytest-4.3.1, py-1.8.0, pluggy-0.9.0
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/Users/enojb/current/sympy/sympy/.hypothesis/examples')
architecture: 64-bit
cache:        no
ground types: python 

rootdir: /Users/enojb/current/sympy/sympy, inifile: pytest.ini
plugins: xdist-1.27.0, instafail-0.4.1, forked-1.0.2, doctestplus-0.3.0, cov-2.7.1, hypothesis-4.32.3
collected 49 items / 47 deselected / 2 selected                                                                                                              

sympy/vector/tests/test_dyadic.py F

________________________________________________________________________ test_dyadic _________________________________________________________________________
sympy/vector/tests/test_dyadic.py:65: in test_dyadic
    assert express(d1, B) == ((cos(q)**2) * (B.i | B.i) + (-sin(q) * cos(q)) *
E   assert (cos(q)**2)*(B.i|B.i) + (-sin(q)*cos(q))*(B.i|B.j) + (-sin(q)*cos(q))*(B.j|B.i) + (sin(q)**2)*(B.j|B.j) == (((((cos(q) ** 2) * (B.i | B.i)) + ((-sin(q) * cos(q)) * (B.i | B.j))) + ((-sin(q) * cos(q)) * (B.j | B.i))) + ((sin(q) ** 2) * (B.j | B.j)))
E    +  where (cos(q)**2)*(B.i|B.i) + (-sin(q)*cos(q))*(B.i|B.j) + (-sin(q)*cos(q))*(B.j|B.i) + (sin(q)**2)*(B.j|B.j) = express((A.i|A.i), B)
E    +  and   cos(q) = cos(q)
E    +  and   B.i = B.i
E    +  and   B.i = B.i
E    +  and   sin(q) = sin(q)
E    +  and   cos(q) = cos(q)
E    +  and   B.i = B.i
E    +  and   B.j = B.j
E    +  and   sin(q) = sin(q)
E    +  and   cos(q) = cos(q)
E    +  and   B.j = B.j
E    +  and   B.i = B.i
E    +  and   sin(q) = sin(q)
E    +  and   B.j = B.j
E    +  and   B.j = B.j

sympy/vector/tests/test_dyadic.py .                                                                                                                    [100%]

                                                                       DO *NOT* COMMIT!                                                                       
===================================================== 1 failed, 1 passed, 47 deselected in 3.65 seconds ======================================================
```

--- a/pytest.ini
+++ b/pytest.ini
@@ -23,3 +23,6 @@ norecursedirs =
     sympy/integrals/rubi/rubi_tests
 
 doctestplus = enabled
+
+markers =
+    nocache_fail

--- a/sympy/calculus/tests/test_util.py
+++ b/sympy/calculus/tests/test_util.py
@@ -155,7 +155,7 @@ def test_periodicity():
         cos, sin, sec, csc, tan, cot))
     assert periodicity(Abs(sin(tan(x))), x) == pi
     assert periodicity(Abs(sin(sin(x) + tan(x))), x) == 2*pi
-    assert periodicity(sin(x) > S.Half, x) is 2*pi
+    assert periodicity(sin(x) > S.Half, x) == 2*pi
 
     assert periodicity(x > 2, x) is None
     assert periodicity(x**3 - x**2 + 1, x) is None

--- a/sympy/diffgeom/diffgeom.py
+++ b/sympy/diffgeom/diffgeom.py
@@ -1036,8 +1036,8 @@ class BaseCovarDerivativeOp(Expr):
                       for k in range(v._coord_sys.dim)])
             derivs.append(d)
         to_subs = [wrt_vector(d) for d in d_funcs]
-        # This substitution can fail when there are Dummy symbols and the
-        # cache is disabled:
+        # XXX: This substitution can fail when there are Dummy symbols and the
+        # cache is disabled: https://github.com/sympy/sympy/issues/17794
         result = d_result.subs(list(zip(to_subs, derivs)))
 
         # Remove the dummies

--- a/sympy/diffgeom/diffgeom.py
+++ b/sympy/diffgeom/diffgeom.py
@@ -1036,6 +1036,8 @@ class BaseCovarDerivativeOp(Expr):
                       for k in range(v._coord_sys.dim)])
             derivs.append(d)
         to_subs = [wrt_vector(d) for d in d_funcs]
+        # This substitution can fail when there are Dummy symbols and the
+        # cache is disabled:
         result = d_result.subs(list(zip(to_subs, derivs)))
 
         # Remove the dummies

--- a/sympy/diffgeom/tests/test_diffgeom.py
+++ b/sympy/diffgeom/tests/test_diffgeom.py
@@ -8,7 +8,7 @@ from sympy.core import Symbol, symbols
 from sympy.simplify import trigsimp, simplify
 from sympy.functions import sqrt, atan2, sin
 from sympy.matrices import Matrix
-from sympy.utilities.pytest import raises
+from sympy.utilities.pytest import raises, nocache_fail
 
 TP = TensorProduct
 
@@ -110,6 +110,7 @@ def test_lie_derivative():
         R2.e_x, TensorProduct(R2.dx, R2.dy))(R2.e_x, R2.e_y) == 0
 
 
+@nocache_fail
 def test_covar_deriv():
     ch = metric_to_Christoffel_2nd(TP(R2.dx, R2.dx) + TP(R2.dy, R2.dy))
     cvd = BaseCovarDerivativeOp(R2_r, 0, ch)

--- a/sympy/diffgeom/tests/test_diffgeom.py
+++ b/sympy/diffgeom/tests/test_diffgeom.py
@@ -114,6 +114,7 @@ def test_covar_deriv():
     ch = metric_to_Christoffel_2nd(TP(R2.dx, R2.dx) + TP(R2.dy, R2.dy))
     cvd = BaseCovarDerivativeOp(R2_r, 0, ch)
     assert cvd(R2.x) == 1
+    # This line fails if the cache is disabled:
     assert cvd(R2.x*R2.e_x) == R2.e_x
     cvd = CovarDerivativeOp(R2.x*R2.e_x, ch)
     assert cvd(R2.x) == R2.x

--- a/sympy/functions/combinatorial/tests/test_comb_numbers.py
+++ b/sympy/functions/combinatorial/tests/test_comb_numbers.py
@@ -439,6 +439,7 @@ def test_genocchi():
     raises(ValueError, lambda: genocchi(-2))
 
 
+@nocache_fail
 def test_partition():
     partition_nums = [1, 1, 2, 3, 5, 7, 11, 15, 22]
     for n, p in enumerate(partition_nums):

--- a/sympy/functions/combinatorial/tests/test_comb_numbers.py
+++ b/sympy/functions/combinatorial/tests/test_comb_numbers.py
@@ -167,6 +167,8 @@ def test_bell():
     # For large numbers, this is too slow
     # For nonintegers, there are significant precision errors
     for i in [0, 2, 3, 7, 13, 42, 55]:
+        # Running without the cache this is either very slow or goes into an
+        # infinite loop.
         assert bell(i).evalf() == bell(n).rewrite(Sum).evalf(subs={n: i})
 
     m = Symbol("m")

--- a/sympy/functions/combinatorial/tests/test_comb_numbers.py
+++ b/sympy/functions/combinatorial/tests/test_comb_numbers.py
@@ -14,7 +14,7 @@ from sympy.core.compatibility import range
 from sympy.core.expr import unchanged
 from sympy.core.numbers import GoldenRatio, Integer
 
-from sympy.utilities.pytest import XFAIL, raises
+from sympy.utilities.pytest import XFAIL, raises, nocache_fail
 
 
 x = Symbol('x')
@@ -131,6 +131,7 @@ def test_tribonacci():
     raises(ValueError, lambda: tribonacci(-1, x))
 
 
+@nocache_fail
 def test_bell():
     assert [bell(n) for n in range(8)] == [1, 1, 2, 5, 15, 52, 203, 877]
 

--- a/sympy/series/tests/test_limits.py
+++ b/sympy/series/tests/test_limits.py
@@ -12,7 +12,7 @@ from sympy.core.add import Add
 from sympy.core.mul import Mul
 from sympy.series.limits import heuristics
 from sympy.series.order import Order
-from sympy.utilities.pytest import XFAIL, raises
+from sympy.utilities.pytest import XFAIL, raises, nocache_fail
 
 from sympy.abc import x, y, z, k
 n = Symbol('n', integer=True, positive=True)
@@ -547,6 +547,7 @@ def test_issue_15984():
     assert limit((-x + log(exp(x) + 1))/x, x, oo, dir='-').doit() == 0
 
 
+@nocache_fail
 def test_issue_13575():
     # This fails with infinite recursion when run without the cache:
     result = limit(acos(erfi(x)), x, 1)

--- a/sympy/series/tests/test_limits.py
+++ b/sympy/series/tests/test_limits.py
@@ -548,6 +548,7 @@ def test_issue_15984():
 
 
 def test_issue_13575():
+    # This fails with infinite recursion when run without the cache:
     result = limit(acos(erfi(x)), x, 1)
     assert isinstance(result, Add)
 

--- a/sympy/solvers/tests/test_solveset.py
+++ b/sympy/solvers/tests/test_solveset.py
@@ -791,6 +791,7 @@ def test_solve_trig():
 
     assert solveset_real(sin(x)**2 + cos(x)**2, x) == S.EmptySet
 
+    # This fails when running with the cache off:
     assert solveset_complex(cos(x) - S.Half, x) == \
         Union(imageset(Lambda(n, 2*n*pi + pi*Rational(5, 3)), S.Integers),
               imageset(Lambda(n, 2*n*pi + pi/3), S.Integers))

--- a/sympy/solvers/tests/test_solveset.py
+++ b/sympy/solvers/tests/test_solveset.py
@@ -30,7 +30,8 @@ from sympy.sets.sets import (Complement, EmptySet, FiniteSet,
 from sympy.tensor.indexed import Indexed
 from sympy.utilities.iterables import numbered_symbols
 
-from sympy.utilities.pytest import XFAIL, raises, skip, slow, SKIP
+from sympy.utilities.pytest import (XFAIL, raises, skip, slow, SKIP,
+    nocache_fail)
 from sympy.utilities.randtest import verify_numerically as tn
 from sympy.physics.units import cm
 
@@ -772,6 +773,7 @@ def test_solveset_complex_tan():
         imageset(Lambda(n, pi*n + pi/2), S.Integers)
 
 
+@nocache_fail
 def test_solve_trig():
     from sympy.abc import n
     assert solveset_real(sin(x), x) == \

--- a/sympy/stats/frv_types.py
+++ b/sympy/stats/frv_types.py
@@ -482,6 +482,7 @@ class HypergeometricDistribution(SingleFiniteDistribution):
         return S(binomial(m, k) * binomial(N - m, n - k))/binomial(N, n)
 
     def _sample_scipy(self, size):
+        import scipy.stats # Make sure that stats is imported
         N, m, n = int(self.N), int(self.m), int(self.n)
         return scipy.stats.hypergeom.rvs(M=m, n=n, N=N, size=size)
 

--- a/sympy/stats/tests/test_discrete_rv.py
+++ b/sympy/stats/tests/test_discrete_rv.py
@@ -11,7 +11,7 @@ from sympy.stats.drv_types import (PoissonDistribution, GeometricDistribution,
                                    Poisson, Geometric, Logarithmic, NegativeBinomial, Skellam,
                                    YuleSimon, Zeta)
 from sympy.stats.rv import sample
-from sympy.utilities.pytest import slow, raises
+from sympy.utilities.pytest import slow, nocache_fail
 
 x = Symbol('x')
 
@@ -51,6 +51,7 @@ def test_Logarithmic():
     assert isinstance(E(x, evaluate=False), Sum)
 
 
+@nocache_fail
 def test_negative_binomial():
     r = 5
     p = S.One / 3

--- a/sympy/stats/tests/test_discrete_rv.py
+++ b/sympy/stats/tests/test_discrete_rv.py
@@ -56,6 +56,7 @@ def test_negative_binomial():
     p = S.One / 3
     x = NegativeBinomial('x', r, p)
     assert E(x) == p*r / (1-p)
+    # This hangs when run with the cache disabled:
     assert variance(x) == p*r / (1-p)**2
     assert E(x**5 + 2*x + 3) == Rational(9207, 4)
     assert isinstance(E(x, evaluate=False), Sum)

--- a/sympy/utilities/pytest.py
+++ b/sympy/utilities/pytest.py
@@ -158,6 +158,10 @@ if not USE_PYTEST:
         func_wrapper.__wrapped__ = func
         return func_wrapper
 
+    def nocache_fail(func):
+        "Dummy decorator for marking tests that fail when cache is disabled"
+        return func
+
     @contextlib.contextmanager
     def warns(warningcls, **kwargs):
         '''Like raises but tests that warnings are emitted.
@@ -200,6 +204,7 @@ else:
     XFAIL = py.test.mark.xfail
     SKIP = py.test.mark.skip
     slow = py.test.mark.slow
+    nocache_fail = py.test.mark.nocache_fail
 
 
 @contextlib.contextmanager

--- a/sympy/utilities/tests/test_wester.py
+++ b/sympy/utilities/tests/test_wester.py
@@ -935,6 +935,7 @@ def test_M14():
     assert solveset_real(tan(x) - 1, x) == ImageSet(Lambda(n, n*pi + pi/4), S.Integers)
 
 
+# This test fails when running with the cache off:
 def test_M15():
     if PY3:
         n = Dummy('n')

--- a/sympy/utilities/tests/test_wester.py
+++ b/sympy/utilities/tests/test_wester.py
@@ -27,7 +27,8 @@ from sympy.functions.combinatorial.numbers import stirling
 from sympy.functions.special.delta_functions import Heaviside
 from sympy.functions.special.error_functions import Ci, Si, erf
 from sympy.functions.special.zeta_functions import zeta
-from sympy.utilities.pytest import XFAIL, slow, SKIP, skip, ON_TRAVIS
+from sympy.utilities.pytest import (XFAIL, slow, SKIP, skip, ON_TRAVIS,
+    raises, nocache_fail)
 from sympy.utilities.iterables import partitions
 from mpmath import mpi, mpc
 from sympy.matrices import Matrix, GramSchmidt, eye
@@ -935,10 +936,11 @@ def test_M14():
     assert solveset_real(tan(x) - 1, x) == ImageSet(Lambda(n, n*pi + pi/4), S.Integers)
 
 
-# This test fails when running with the cache off:
+@nocache_fail
 def test_M15():
     if PY3:
         n = Dummy('n')
+        # This test fails when running with the cache off:
         assert solveset(sin(x) - S.Half) in (Union(ImageSet(Lambda(n, 2*n*pi + pi/6), S.Integers),
                                                ImageSet(Lambda(n, 2*n*pi + pi*R(5, 6)), S.Integers)),
                                                Union(ImageSet(Lambda(n, 2*n*pi + pi*R(5, 6)), S.Integers),
@@ -1334,7 +1336,6 @@ def test_P3():
     A222 = -A[(3, 0), (2, 1)]
     A22 = BlockMatrix([[A221, A222]]).T
     rows = [[-A11, A12], [A21, A22]]
-    from sympy.utilities.pytest import raises
     raises(ValueError, lambda: BlockMatrix(rows))
     B = Matrix(rows)
     assert B == Matrix([

--- a/sympy/vector/tests/test_dyadic.py
+++ b/sympy/vector/tests/test_dyadic.py
@@ -4,10 +4,12 @@ from sympy.vector import (CoordSys3D, Vector, Dyadic,
                           DyadicAdd, DyadicMul, DyadicZero,
                           BaseDyadic, express)
 
+from sympy.utilities.pytest import nocache_fail
 
 A = CoordSys3D('A')
 
 
+@nocache_fail
 def test_dyadic():
     a, b = symbols('a, b')
     assert Dyadic.zero != 0

--- a/sympy/vector/tests/test_dyadic.py
+++ b/sympy/vector/tests/test_dyadic.py
@@ -62,6 +62,7 @@ def test_dyadic():
     q = symbols('q')
     B = A.orient_new_axis('B', q, A.k)
     assert express(d1, B) == express(d1, B, B)
+    # This assertion fails when running with the cache off:
     assert express(d1, B) == ((cos(q)**2) * (B.i | B.i) + (-sin(q) * cos(q)) *
             (B.i | B.j) + (-sin(q) * cos(q)) * (B.j | B.i) + (sin(q)**2) *
             (B.j | B.j))

